src/map/pc.cpp

```diff
@@ -12143,253 +12218,466 @@ static unsigned int pc_calc_basesp(uint16 level, uint16 class_) {
-	}
-//	ShowInfo("%s - Class %d: %d\n", type?"Job":"Base", job_id, job_info[idx].max_level[type]);
-	for (i = 1; i < job_count; i++) {
-		job_id = jobs[i];
-		if (!pcdb_checkid(job_id)) {
-			ShowError("pc_readdb_job_exp: Invalid job ID %d.\n", job_id);
-			continue;
+					job->aspd_base[static_cast<int16>(constant)] = aspd;
+				}
+			}
+
+			if (this->nodeExists(node, "BonusStats")) {
+				const YAML::Node &bonusNode = node["BonusStats"];
+				job->job_bonus.resize(MAX_LEVEL);
+
+				for (const YAML::Node &levelNode : bonusNode) {
+					uint16 level;
+
+					if (!this->asUInt16(levelNode, "Level", level))
+						return 0;
+
+					if (level > MAX_LEVEL) {
+						this->invalidWarning(levelNode["Level"], "Level must be between 1~MAX_LEVEL for %s.\n", job_name.c_str());
+						return 0;
+					}
+
+					for (uint8 idx = PARAM_STR; idx < PARAM_MAX; idx++) {
+						if (this->nodeExists(levelNode, parameter_names[idx])) {
+							int16 change;
+
+							if (!this->asInt16(levelNode, parameter_names[idx], change))
+								return 0;
+
+							job->job_bonus[level - 1][idx] = change;
+						}
+					}
+				}
+			}
+
+			if (this->nodeExists(node, "MaxStats")) {
+				const YAML::Node &statNode = node["MaxStats"];
+
+				for (const auto &statit : statNode) {
+					std::string stat = statit.first.as<std::string>(), stat_constant = "PARAM_" + stat;
+					int64 constant;
+
+					if (!script_get_constant(stat_constant.c_str(), &constant)) {
+						this->invalidWarning(statNode["Bonus"], "Unknown max stat %s specified for %s, skipping.\n", stat.c_str(), job_name.c_str());
+						continue;
+					}
+
+					if (constant < PARAM_STR || constant >= PARAM_MAX) {
+						this->invalidWarning(statNode["Bonus"], "Invalid max stat %s specified for %s, skipping.\n", stat.c_str(), job_name.c_str());
+						continue;
+					}
+
+					uint16 max;
+
+					if (!this->asUInt16(statNode, stat.c_str(), max))
+						return 0;
+
+					job->max_param[constant] = max;
+				}
+			}
+
+			if (this->nodeExists(node, "MaxBaseLevel")) {
+				uint16 level;
+
+				if (!this->asUInt16(node, "MaxBaseLevel", level))
+					return 0;
+
+				if (level > MAX_LEVEL) {
+					this->invalidWarning(node["MaxBaseLevel"], "MaxBaseLevel must be between 1~MAX_LEVEL for %s, capping to MAX_LEVEL.\n", job_name.c_str());
+					level = MAX_LEVEL;
+				}
+
+				job->max_base_level = level;
+			} else {
+				if (!exists)
+					job->max_base_level = MAX_LEVEL;
+			}
+
+			if (this->nodeExists(node, "BaseExp")) {
+				for (const YAML::Node &bexpNode : node["BaseExp"]) {
+					uint16 level;
+
+					if (!this->asUInt16(bexpNode, "Level", level))
+						return 0;
+
+					if (level < 1 || level > MAX_LEVEL) {
+						this->invalidWarning(bexpNode["Level"], "Level must be between 1~MAX_LEVEL for %s.\n", job_name.c_str());
+						return 0;
+					}
+
+					if (this->nodeExists(bexpNode, "Exp")) {
+						uint64 exp;
+
+						if (!this->asUInt64(bexpNode, "Exp", exp))
+							return 0;
+
+						job->base_exp[level - 1] = exp;
```

Comment:
Maybe as `t_exp`, for older client?
```
/// Max Base and Job EXP for players
#if PACKETVER >= 20170830
	const t_exp MAX_EXP = INT64_MAX;
#else
	const t_exp MAX_EXP = INT32_MAX;
#endif
```

Reply:
You mean checking the limit?
Serverside we always store it as int64 now, but we sometimes cap it to int32 for older client versions.